cmd/promtool/main.go

```diff
@@ -303,9 +304,74 @@ func checkRules(filename string) (int, []error) {
 		numRules += len(rg.Rules)
 	}
 
+	showWarning := false
+	dLabels, dMetrics, nLabels, nMetrics := checkDuplicates(rgs.Groups)
+	if dLabels != 0 {
+		showWarning = true
+		fmt.Printf("%d duplicate label(s) found.\n", dLabels)
+		for _, n := range nLabels {
+			fmt.Println(n)
+		}
+	}
+	if dMetrics != 0 {
+		showWarning = true
+		fmt.Printf("%d duplicate metric(s) found.\n", dMetrics)
+		for _, n := range nMetrics {
+			fmt.Println(n)
+		}
+	}
+	if showWarning {
+		fmt.Println("Might cause inconsistency while recording expressions.")
+	}
+
 	return numRules, nil
 }
 
+func checkDuplicates(r []rulefmt.RuleGroup) (int, int, []string, []string) {
+	duplicateLabels, duplicateMetrics := 0, 0
+	labelNames, metricName := []string{}, []string{}
+	for _, rule := range r {
+		var stackLabels []map[string]string
+		var stackMetrics []string
+		for _, props := range rule.Rules {
+			for _, labels := range stackLabels {
+				if reflect.DeepEqual(labels, props.Labels) {
+					duplicateLabels++
+					found := false
+					for _, inst := range labelNames {
+						if inst == labels["label"] {
+							found = true
+							break
+						}
+					}
+					if !found {
+						labelNames = append(labelNames, labels["label"])
+					}
+				}
+			}
+			stackLabels = append(stackLabels, props.Labels)
+
+			for _, mname := range stackMetrics {
```

Comment:
You need to check the metric names and labels at the same time, not independently.

Reply:
sure